keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,1169 @@
+		defer close(clientDoneCh)
+		client.Get(server.URL + "/foo")
+	}()
+
+	select {
+	case <-handlerDoneCh:
+	case <-time.After(wait.ForeverTestTimeout):
+		t.Errorf("expected the request handler to have terminated")
+	}
+}
+```
+
+Even though the read/write deadline is set on the server side for this request, the
+client never receives any response or error from the server. We have opened an
+[issue](https://github.com/golang/go/issues/65526) with the golang team to track this.
+
+
+With the removal of the `timeout` filter, we need to ensure that a client, not 
+enforcing any client-side timeout, never blocks indefinitely due to the corresponding
+request handler being frozen on the server side. In order to maintain parity, the
+client should receive a response/error after the allotted deadline for the 
+request elapses on the server.
+
+
+Mitigation Steps:
+- a) If this is a bug, we can wait until the fix is in the net/http package.
+- b) We can enable this feature for http/2.0 request(s) only, all
+     http/1x request(s) will continue to be served using the `timeout` filter.
+- c) We enable this feature for both http/1x and http/2.0; for http/1x requests
+     we keep the `timeout` filter enabled.
+
+Unit tests must be in place to verify that a frozen request handler
+does not, as a consequence, block the corresponding client:
+```
+neverDoneCh := make(chan struct{})
+func(w http.ResponseWriter, req *http.Request) {
+	<-neverDoneCh
+}
+```
+
+
+##### Request Handler Running Indefinitely on the Server:
+
+Since there is no mechanism to stop a goroutine from the outside, we rely on the
+request handler goroutine to self-terminate as soon as possible once the read/write 
+deadline exceeds. We provide the following mechanisms to the handler so it knows 
+when to terminate:
+- The handler detects that the `context` associated with the request has exceeded
+  its deadline, and so it terminates.
+- In the course of serving a request, the handler invokes the `Write` or `FlushError`
+  method of the `ResponseWriter` object. The invocation will result in either an error
+  or a panic once the read/write deadline exceeds. The handler terminates as a consequence
+
+
+The proposed feature will be in par with the `timeout` filter, if we decorate the
+`ResponseWriter` object to throw a panic once the read/write deadline exceeds.
+
+
+##### Impact When an HTTP/1x Connection is Hijacked:
+
+The hijacker is responsible for managing the state of a hijacked connection. Add
+a test to demonstrate that setting a read/write deadline before
+the connection is hijacked does not interfere with the operation of the hijacker.
+
+There are two code sites that invoke `Hijack`:
+- apimachinery/pkg/util/proxy/upgradeaware.go
+- apimachinery/pkg/util/httpstream/spdy/upgrade.go
+
+Add test(s) to verify that with read/write timeout enabled, the above 
+features work as expected with hijacked connections.
+
+## Design Details
+
+### How We Manage Request Timeout Today:
```

Comment:
How similar is the Kubernetes timeout filter to https://pkg.go.dev/net/http#TimeoutHandler? Are there examples of attempts outside of Kubernetes to address similar concerns?